src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -394,4 +355,145 @@ private[concurrent] final object Promise {
+      case a: Callback[U] =>
+        if (a eq NoopCallback) this
+        else if (this eq NoopCallback) a
+        else if (this eq a) this
+        else ManyCallbacks.two(this, a)
+      case m: ManyCallbacks[U] => m append this // m append this == this prepend m
+    }
+
+    override def toString: String = s"Callback($executor, $onComplete)"
+  }
+
+  object ManyCallbacks {
+    def two[U](second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](NoopCallback, NoopCallback, second, first, 2)
+    def three[U](third: Callbacks[U], second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](NoopCallback, third, second, first, 1)
+    def four[U](fourth: Callbacks[U], third: Callbacks[U], second: Callbacks[U], first: Callbacks[U]): ManyCallbacks[U] =
+      new ManyCallbacks[U](fourth, third, second, first, 0)
+  }
+
+  final class ManyCallbacks[+T] private[ManyCallbacks] (
+    final val c1: Callbacks[T],
+    final val c2: Callbacks[T],
+    final val c3: Callbacks[T],
+    final val c4: Callbacks[T],
+    private[ManyCallbacks] final val remainingCapacity: Int) extends Callbacks[T] {
+
+    //Don't want to incur the runtime ovehead of these checks, but this invariant will hold true:
+    //require(c3 ne NoopCallback)
+    //require(c4 ne NoopCallback)
+
+    private[this] final def merge0[U >: T](m: ManyCallbacks[U]): ManyCallbacks[U] =
+      (m.remainingCapacity: @switch) match {
+        case 0 => ManyCallbacks.two(this, m)
+        case 1 => ManyCallbacks.four(c1, c2, c3, ManyCallbacks.four(c4, m.c2, m.c3, m.c4))
+        case 2 => ManyCallbacks.three(c1, c2, ManyCallbacks.four(c3, c4, m.c3, m.c4))
+      }
+
+    private[this] final def merge1[U >: T](m: ManyCallbacks[U]): ManyCallbacks[U] =
+      (m.remainingCapacity: @switch) match {
+        case 0 => ManyCallbacks.four(c2, c3, c4, m)
+        case 1 => ManyCallbacks.three(c2, c3, ManyCallbacks.four(c4, m.c2, m.c3, m.c4))
+        case 2 => ManyCallbacks.two(c2, ManyCallbacks.four(c3, c4, m.c3, m.c4))
+      }
+
+    private[this] final def merge2[U >: T](m: ManyCallbacks[U]): ManyCallbacks[U] =
+      (m.remainingCapacity: @switch) match {
+        case 0 => ManyCallbacks.four(c2, c3, c4, m)
+        case 1 => ManyCallbacks.two(c3, ManyCallbacks.four(c4, m.c2, m.c3, m.c4))
+        case 2 => ManyCallbacks.four(c3, c4, m.c3, m.c4)
+      }
+
+    private[ManyCallbacks] final def merge[U >: T](m: ManyCallbacks[U]): ManyCallbacks[U] =
+      if (this ne m) { // Do not merge with itself
+        (remainingCapacity: @switch) match {
+          case 0 => merge0(m)
+          case 1 => merge1(m)
+          case 2 => merge2(m)
+        }
+      } else this
+
+    final def append[U >: T](c: Callbacks[U]): Callbacks[U] = c match {
+      case m: ManyCallbacks[U]    => this merge m
+      case a if a eq NoopCallback => this // Don't append Noops
+      case a: Callback[U]         =>
+        (remainingCapacity: @switch) match {
```

Comment:
How about merging the two switch levels into a single one, where you switch on `(remainingCapacity << 2) | m.remainingCapacity`?